### ISSUE TITLE
GCCollector / multiprocess mode deadlock

### ISSUE THREAD

**Issue by akx** [prometheus/client_python#322]
Ref #313, #321. 

There's a chance for a deadlock with the GCCollector and multiprocess mode (when `prometheus_multiproc_dir` is set).

For instance, when running our test suite with `py.test`, the process seems to hang, and ctrl+c yields this (on 0.4.0):

```
Traceback (most recent call last):
  File "prometheus_client/gc_collector.py", line 49, in _cb
    latency.labels(gen).observe(delta)
  File "prometheus_client/core.py", line 747, in labels
    self._metrics[labelvalues] = self._wrappedClass(self._name, self._labelnames, labelvalues, **self._kwargs)
  File "prometheus_client/core.py", line 1088, in __init__
    self._sum = _ValueClass(self._type, name, name + '_sum', labelnames, labelvalues)
  File "prometheus_client/core.py", line 636, in __init__
    with lock:
```

This looks like a problem with GC collector callbacks firing while another metric's value object (a `_MultiProcessValue`) is being modified (which holds the `Lock()` shared by all values created [here](https://github.com/prometheus/client_python/blob/9e65e887694137bfa53fef3667fbdde0ba7b0772/prometheus_client/core.py#L623)).

One possible option might be to use an `RLock()` instead of a `Lock()`, but I'm not sure what that might cause.

**Comment by akx** [comment]
@tcolgate See here: https://github.com/prometheus/client_python#multiprocess-mode-gunicorn

The crux of the issue is that in multiprocess mode there's an additional lock that prevents concurrent/reentrant writes of values (since they're backed by a file, not bare Python memory).

Since the GC callback may occur at any time, and Python runs the callback code in the main thread, there's a nontrivial chance that the GC callback is called when something else is updating a metric value, causing a deadlock.

**Comment by tcolgate** [comment]
Why is that a deadlock rather than just a lock contention?
Can multiproc be detected? I'd suggest skipping the GC collector if so.

**Comment by akx** [comment]
> Why is that a deadlock rather than just a lock contention?

Because the GC callbacks [are run as "interrupts"](https://github.com/python/cpython/blob/master/Modules/gcmodule.c#L1153-L1190) in the same thread that happens to make an allocation that causes garbage collection, so you may get this happening in a single thread:

1. metric value is updated
2. acquire lock for metric writing
3. allocation occurs for something within metric writing, causes GC
4. GC callback is called, updates a metric
5. deadlock attempting to reacquire same lock we already hold

**Comment by akx** [comment]
Adding more threads into the mix sounds unnecessary. I think queue processing could be done on-demand, at `collect()` time for the `GCCollector`.

**Comment by tcolgate** [comment]
Possibly, though you wouldn't want an unbounded queue, and if a queue fills
(because prom has been down for a while), the resulting scrapes will be
given incorrect info. Personally I don't think another thread is a major
issue, but I don't really understand python threading (or python), or this
multiproc stuff well enough.

**Comment by charan28** [comment]
@akx I'm coming across this issue in 0.4.2, without `prometheus_multiproc_dir` set. Reading through this bug and the associated fix, it seems like this behavior was only noticed when that environment variable was set - so unsure what's causing this.

Relevant failure:
```
Exception ignored in: <function GCCollector.__init__.<locals>._cb at 0x7fe8dd6b1488>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/prometheus_client/gc_collector.py", line 58, in _cb
    latency.labels(gen).observe(delta)
  File "/usr/lib/python3.6/site-packages/prometheus_client/core.py", line 777, in labels
    labelvalues = tuple(unicode(l) for l in labelvalues)
  File "/usr/lib/python3.6/site-packages/prometheus_client/core.py", line 777, in <genexpr>
    labelvalues = tuple(unicode(l) for l in labelvalues)
RecursionError: maximum recursion depth exceeded while getting the str of an object
Fatal Python error: Cannot recover from stack overflow.
```

Any suggestions on what I might be doing wrong?

**Comment by akx** [comment]
Ugh, that sounds bad.

An educated guess is that a GC collection can occur even within the same thread the GC collection callback is running in, so the callback gets called again, and... well, you can guess the recursive rest.

The quickest fix (if the above guess is correct) is to add a re-entrancy lock to the GC callback function, so it won't do anything if it's already running. I'm not sure if it's safe to do so within the GC callback, but adding a pair of  `gc.disable()` and `gc.enable()` should also work.

On the bright side: You're not doing anything wrong there, @charan28.

**Comment by akx** [comment]
@charan28 Assuming my hypothesis is right, the above ☝️ PR should fix this situation.

**Comment by xudifsd** [comment]
We have encountered this deadlock in production environment, the python stack trace is:
```
Current thread 0x00007f1a14b31700 (most recent call first):
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/metrics.py", line 148 in labels
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/gc_collector.py", line 72 in _cb
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/metrics.py", line 179 in _multi_samples
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/metrics.py", line 68 in collect
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/registry.py", line 75 in collect
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/exposition.py", line 89 in generate_latest
  File "/usr/local/lib/python3.7/site-packages/prometheus_client/exposition.py", line 45 in prometheus_app
  File "/usr/local/lib/python3.7/wsgiref/handlers.py", line 137 in run
  File "/usr/local/lib/python3.7/wsgiref/simple_server.py", line 133 in handle
  File "/usr/local/lib/python3.7/socketserver.py", line 717 in __init__
  File "/usr/local/lib/python3.7/socketserver.py", line 357 in finish_request
  File "/usr/local/lib/python3.7/socketserver.py", line 344 in process_request
  File "/usr/local/lib/python3.7/socketserver.py", line 313 in _handle_request_noblock
  File "/usr/local/lib/python3.7/socketserver.py", line 234 in serve_forever
  File "/job_exporter/main.py", line 129 in main
  File "/job_exporter/main.py", line 162 in <module>
```

You can see C stack trace [here](https://github.com/Microsoft/pai/issues/2104). Our code is [here](https://github.com/Microsoft/pai/blob/dixu/debug-deadlock/src/job-exporter/src/main.py#L125). We are using the latest code from [pip3](https://github.com/Microsoft/pai/blob/dixu/debug-deadlock/src/job-exporter/build/job-exporter.dockerfile#L33). I know this issue from the comment in code.

Still looking what happened.

**Comment by xudifsd** [comment]
It seems [this](https://github.com/prometheus/client_python/blob/master/prometheus_client/metrics.py#L178) and [this](https://github.com/prometheus/client_python/blob/master/prometheus_client/metrics.py#L148) locks the same lock. Because GC can interrupt call to metrics._multi_samples and during gc callback it will call metrics.labels. It should be RLock.

**Comment by brian-brazil** [comment]
Fixed by #371
